docs(design): improve design chapter (#139)
tier4/CARET_doc: docs/design/software_architecture/index.md

![architecture](../../imgs/architecture.drawio.png)

CARET serves three phases; recording, configuration and visualization.

## Recording

In the recording phase, CARET records events data obtained from tracepoints during application's runtime.

> CARET adopts LTTng as a tracing mechanism. LTTng session daemon collects events from tracepoints embedded in a target application.

Reviewer: ```suggestion
CARET adopts LTTng as a tracing mechanism. LTTng session daemon collects events from tracepoints.
```

`rclcpp` provided by ROS 2 has original tracepoints and CARET uses some of them also. CARET collects its dedicated tracepoints added by function hooking.

> To add tracepoints flexibly, function hooking is .

Reviewer: ```suggestion
Function hooks are actively used in CARET because they do not require changes to multiple existing packages and allow adding trace points in a flexible way.
```

Author: I'd like to align `function hooking`, instead of `hooks` or `function hooks`.

Only if it is not possible to add tracepoints by function hooking due to some constraints from implementation, tracepoints are added via another approach.

> All of recorded events are stored to a set of trace data CTF-based. It is visualized for user to observe application's performance and behavior.

Reviewer: ```suggestion
All of recorded events are stored to a set of trace data CTF-based. It is visualized for users to observe application's performance and behavior.
```

[`caret_trace`](./caret_trace.md) is the main package for realizing recording. `caret_trace` collects events invoked in `rclcpp`, `rcl` and DDS. It is inconvenient to see actual time when data is consumed because data is consumed in a user code actually. [TILDE](./tilde.md) serves tracepoints to collect events which happens in a user code. CARET is able to refer to them for diving into events in a user code.

See also

- [Tracepoints](../trace_points/index.md)
- [Runtime Processing](../runtime_processing/index.md)
- [The LTTng Documentation](https://lttng.org/docs/)

## Configuration

In the configuration, CARET expects users to complement structure definitions of a target application before visualizing data. Politely speaking, CARET expects users to define intra-node data paths and inter-node data paths to calculate response time.

> Extracting such definitions mechanically is difficult because this step requires structure of a target application.

Reviewer: ```suggestion
It is difficult to extracting such definitions mechanically fom trace data because this step depends on implementation of a target application which is not able to grasp from ros-layer.
```

Author: I think that grammar on the suggested sentence is not good, but I'd like to change it with seeing your feedback.

Author: I think that this sentence should be more abstract due to the following reason.
CARET ask users to define inter-path data path because CARET does not know what to observe at first. Users has to define intra-node data path because CARET want to complement information to grasp relationship between sub and pub, which you pointed out.

There are a few reasons for defining paths and strict explanation will confuse users.

Reviewer: I see. LGTM

Structure definitions are stored in a object which is instantiated from Architecture class defined by [`caret_analyze`](./caret_analyze.md).

> CARET serves Python API to deal with the object to fulfill configuration step by running script. CARET is able to store he object to a YAML-based file to be reused.

Reviewer: ```suggestion
CARET serves Python API to deal with the object to fulfill configuration step by running script. CARET is able to store the object to a YAML-based file to be reused.
```
